flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala

```diff
@@ -1073,57 +1073,73 @@ class JobManager(
       // execute the recovery/writing the jobGraph into the SubmittedJobGraphStore asynchronously
       // because it is a blocking operation
       future {
-        try {
-          if (isRecovery) {
-            executionGraph.restoreLatestCheckpointedState()
-          }
-          else {
-            val snapshotSettings = jobGraph.getSnapshotSettings
-            if (snapshotSettings != null) {
-              val savepointPath = snapshotSettings.getSavepointPath()
+        val restoreStateSuccess =
+          try {
+            if (isRecovery) {
+              executionGraph.restoreLatestCheckpointedState()
```

Comment:
Yes, it had the same issue as the savepoint restore. If restoring the checkpoint failed, job submission was never ACK'd, but execution was restarted etc. I'm not sure that this is the behaviour we want in the long term, but I think failing the submission is clearer behaviour right now.

Another issue I've noticed is the following: when HA checkpoint recovery fails and the job cannot be scheduled for execution, the job will eventually be removed from ZooKeeper and hence never be recovered again. Let me open an issue for this...